cibuildwheel/platforms/android.py

```diff
@@ -0,0 +1,625 @@
+    if not match:
+        msg = f"invalid Android identifier: '{identifier}'"
+        raise ValueError(msg)
+    major, minor, arch = match.groups()
+    return (f"{major}.{minor}", arch)
+
+
+@dataclass(frozen=True)
+class PythonConfiguration:
+    version: str
+    identifier: str
+    url: str
+
+    @property
+    def arch(self) -> str:
+        return parse_identifier(self.identifier)[1]
+
+
+def all_python_configurations() -> list[PythonConfiguration]:
+    return [PythonConfiguration(**item) for item in resources.read_python_configs("android")]
+
+
+def get_python_configurations(
+    build_selector: BuildSelector, architectures: set[Architecture]
+) -> list[PythonConfiguration]:
+    return [
+        c
+        for c in all_python_configurations()
+        if c.arch in architectures and build_selector(c.identifier)
+    ]
+
+
+def shell_prepared(command: str, *, build_options: BuildOptions, env: dict[str, str]) -> None:
+    shell(
+        prepare_command(command, project=".", package=build_options.package_dir),
+        env=env,
+    )
+
+
+def before_all(options: Options, python_configurations: list[PythonConfiguration]) -> None:
+    before_all_options = options.build_options(python_configurations[0].identifier)
+    if before_all_options.before_all:
+        log.step("Running before_all...")
+        shell_prepared(
+            before_all_options.before_all,
+            build_options=before_all_options,
+            env=before_all_options.environment.as_dictionary(os.environ),
+        )
+
+
+def build(options: Options, tmp_path: Path) -> None:
+    configs = get_python_configurations(
+        options.globals.build_selector, options.globals.architectures
+    )
+    if not configs:
+        return
+
+    try:
+        before_all(options, configs)
+
+        built_wheels: list[Path] = []
+        for config in configs:
+            log.build_start(config.identifier)
+            build_options = options.build_options(config.identifier)
+            build_path = tmp_path / config.identifier
+            build_path.mkdir()
+
+            python_dir = setup_target_python(config, build_path)
+
+            compatible_wheel = find_compatible_wheel(built_wheels, config.identifier)
+            if compatible_wheel:
+                print(
+                    f"\nFound previously built wheel {compatible_wheel.name} that is "
+                    f"compatible with {config.identifier}. Skipping build step..."
+                )
+                repaired_wheel = compatible_wheel
+            else:
+                build_env, android_env = setup_env(config, build_options, build_path, python_dir)
+                before_build(build_options, build_env)
+                built_wheel = build_wheel(build_options, build_path, android_env)
+                repaired_wheel = repair_wheel(
+                    build_options, build_path, build_env, android_env, built_wheel
+                )
```

Comment:
It will make it more verbose, which is a shame, but I think I'd feel better if these functions used keyword args - there are potential errors where e.g. android_env and build_env could get mixed up and the typechecker wouldn't catch it.

Reply:
I've come up with a different solution to this problem: since the same arguments are generally passed together, I've grouped them into a dataclass. This significantly simplifies both the function declarations and the function calls.